### ISSUE TITLE
Архитектура

### ISSUE THREAD

**Issue by sodiqit** [performanceArtist/dex-sniper#1]
Привет!

Начал делать ревью проекта и заметил, что архитектура кажется слишком запутанной. Ты используешь медиатор, но по сути он тут не нужен. У тебя появляется один раздутый класс, который знает все обо всех по всей системе и уже сейчас при таком маленьком количестве функционала - это уже тяжело читать и вникать. Тут медиатор как просто огромной жирный интерцептор + куча асинхронного поведения(даже рест запрос обычный становится тяжело-читаемым)

Я попрошу использовать обычную 3-ех слойную архитектуру, которая характерна для NEST(controller -> service -> repository). То есть прямо из контроллера вызывать сервис, как в стандартных приложениях без лишней абстракции. Этих слоев для данного приложения с лихвой хватит, чтобы удобно его масштабировать

1. Для телеграм бота предлагаю использовать [nestjs-telegraf](https://www.npmjs.com/package/nestjs-telegraf) и разделить один метод подписки на сообщения на несколько маленьких через декоратор  @Command c валидацией входных параметров. (Можно написать и свои декораторы в рамках обучения, тут по желанию)

2. По поводу асинхронного поведения - здесь оно необходимо по большей степени, когда приходит эвент с блокчейна, тогда можно использовать [event-emitter2](https://docs.nestjs.com/techniques/events) и кидать нужный эвент, который будет консюмить уже веб и тг

3. Реализовать stateless систему. Сейчас заметил, что при подписывании на эвенты в сокете - идет запись в память, но нет никакой очистки. То же самое с UserMediator, зачем этот лишний слой абстракции с генерацией кучи инстансов класса для каждого отдельного юзера - не ясно.

После этого можно будет продолжить более детально смотреть.

**Comment by performanceArtist** [comment]
Привет. Медиатор он тут да, больше как эксперимент. Хотел попробовать полностью изолировать клиент(телеграм, веб, etc) от внутреннего функционала(user, subscriptions). И он по идее ни о чем не знает как раз - он полностью абстрактный и не имеет ни одного внешнего импорта. Знают только снаружи о нем. Ну да ладно, убрал. 

1. Добавил. Хотя здесь он по сути оказался бесполезным, т.к. нужен только самый базовый функционал. Парсинг команд он тоже не предоставляет. 
2. Сделал. Хотя imo event-ы - такое себе, очень неявно.
3. Тут реально не особо нужно их(инстансы `UserMediator`) хранить. На самом деле в них и нет состояния. Инстансы нужны чисто для удобства - чтобы прокинуть один раз id и получить весь функционал(в котором любая функция все равно потребует этот id). Не столько класс, сколь частично примененная функция типа `const func = (config: Config) => (param: number) => {...}` Сокеты я накидал очень быстро чисто для демонстрации, поправлю.  
 
Про трехслойную архитектуру немного не понял по применению к telegram. Controller - не прямо что абстракция, он к http привязан. Ну для http добавил, конечно - оно там так и было.

**Comment by sodiqit** [comment]
Да, он абстрактный, но все типы ответов по всей системе все таки знает, то что он их не импортирует, а хранит у себя в типах - как по мне не очень

1. Почему не предоставляет?

```ts
import { Command } from 'nestjs-telegraf';

    @Command('test')
    async test(@Ctx() ctx: Context) {
        if (!ctx.from || !ctx.payload) {
            return;
        }

        ctx.reply('test');
    }
```

2. Да, тут согласен, что не очень явно, но с другой стороны в сервисе понятно будет, что идет эмит определенного эвента, и проблем посмотреть всех консюмеров через ide - нет как по мне

По поводу трехслойной я имел ввиду базовое понятие контроллера. Он же не только http уровня может быть верно? хендлер telegram по своей сути тоже будет контроллером. По сути я имел ввиду точку входа

например, для тг:

```ts
@Update()
export class AppBotCommandHandler {
    private readonly logger = new Logger(AppBotCommandHandler.name);

    constructor(
        private readonly abstractService: AbstractService,
    ) {}

    @Start()
    @CreateRequestContext<AppBotCommandHandler>((t) => {
        if (!t) {
            throw new Error('Not found this context in create request context');
        }
        return t.rep.orm;
    })
    async start(@Ctx() ctx: Context) {
        const name = await this.abstractService.getName(ctx.from.id);
        ctx.reply(`hello ${name}`);
    }
}

```

вот другой пример для обработчика очереди:

```ts
@Injectable()
export class UserHandler {
    constructor(private readonly userGateway: UserGateway) {} //socket server

    @RabbitSubscribe({
        exchange: API_BACK_EXCHANGE,
        routingKey: BalanceEvents.BAlANCE_UPDATE,
        queueOptions: {
            durable: false,
            autoDelete: true,
        },
    })
    async handleUpdateUserBalance(@RabbitPayload() data: BalanceUpdateMessageDto) {
        this.userGateway.notifyAboutUpdateBalance(data.userId, data.balance);
    }
}
```

а вот пример контроллера для эвентов от эмиттера:

```ts

@Injectable()
export class ReferralEventHandler {
    private readonly logger = new Logger(ReferralEventHandler.name);

    constructor(
        private readonly rep: RepositoryService,
        private readonly service: ReferralService,
    ) {}

    @OnEvent(ReferralEvents.EARN_REWARD)
    @CreateRequestContext((t: ReferralEventHandler) => t.rep.orm)
    async onEarnRewardByReferral({ referral, reward, rewardType }: EarnRewardEvent) {
        if (rewardType !== BalanceItemKind.MTGN) {
            return;
        }
        await this.service.updateReferrerBalances(referral, new Decimal(reward));
    }
}

```

все это контроллеры своего "транспортного" уровня, надеюсь я смог понятно объяснить, что я имею ввиду под контроллером

**Comment by performanceArtist** [comment]
1. У меня нет `ctx.payload`, ts ругается. Может убрали в новой версии. Но я имел в виду скорее какой-нибудь стандартный формат, где указываешь тип каждого аргумента - а там поди в `payload` просто `text.split(' ')` вызывано на сообщении. Ну и когда я искал, вроде писали что нет: https://github.com/telegraf/telegraf/discussions/1291 В любом случае, я уже и так написал регексы - мне кажется норм. 

По контроллеру да, я понял - ты про в общем понятие. Просто что с точки зрения nest это не контроллер, т.к. с http не работает(следовательно, нет декоратора `@Controller` и в `controllers` его класть не надо).

**Comment by performanceArtist** [comment]
Есть кстати вопрос. Вот [тут](https://github.com/performanceArtist/dex-sniper/blob/main/src/modules/user/user.service.ts#L248) есть некрасивый запрос. Суть в чем - нужно найти пул, имея два символа токенов. Причем порядок может быть разным - `WETH/USDT` и `USDT/WETH` это один пул, очевидно. Вообще у uniswap есть возможность вычислить адрес пула, имея символы. Но с этим проблема в том, что там на входе нужно передавать fee. И в результате могут быть выданы разные пулы. Поэтому я в целом подумал, что раз все равно все хранится в базе, то в этом случае лучше пусть она будет одним источником истины. Первое, что в голову приходит - сделать какой-нибудь типа запрос к базе с `in` и id-шниками. Но тут как будто усложнение + нужно как-то случай с одинаковыми id отмести. Как бы ты сделал?
